ledger-service/http-json/src/main/scala/com/digitalasset/http/util/ContractStreamStep.scala

```diff
@@ -29,11 +29,13 @@ private[http] sealed abstract class ContractStreamStep[+D, +C] extends Product w
         Txn(toInsertDelete, off)
       case (Acs(_) | Txn(_, _), Txn(ostep, off)) =>
         Txn(toInsertDelete append ostep, off)
-      case (LiveBegin(_), Txn(_, _) | LiveBegin(_)) => o
+      case (LiveBegin(_), Txn(_, _)) => o
       // the following cases should never happen in a real stream; we attempt to
       // provide definitions that make `append` totally associative, anyway
-      case (Acs(_), LiveBegin(LedgerBegin)) => this
-      case (LiveBegin(LedgerBegin), Acs(_)) => o
+      case (Acs(_) | LiveBegin(_), LiveBegin(LedgerBegin)) => this
+      case (LiveBegin(LedgerBegin), Acs(_) | LiveBegin(_)) |
+          (LiveBegin(AbsoluteBookmark(_)), LiveBegin(AbsoluteBookmark(_))) =>
```

Comment:
do we guarantee that `o.offset` > `this.offset` in the case when we append `LiveBegin(AbsoluteBookmark(_))` to another `LiveBegin(AbsoluteBookmark(_))`?

else it should it be pseudo code:
```
case ((LiveBegin(AbsoluteBookmark(x)), LiveBegin(AbsoluteBookmark(y))) if x < y => o
case ((LiveBegin(AbsoluteBookmark(x)), LiveBegin(AbsoluteBookmark(y))) if x>= y => this
```

Reply:
actually I see the comment above :), the proposed pseudo code will break associativity.

Reply:
It might or might not be associative, but keep in mind the real intent of this semigroup is to preserve the offset _that appears later in the stream_, for which always yielding `o` in this scenario is correct. This intent is made explicit:

https://github.com/digital-asset/daml/blob/f9057bea195abb547d6af2f89f8e09776a45f267/ledger-service/http-json/src/test/scala/com/digitalasset/http/util/ContractStreamStepTest.scala#L43-L50